File: standalone-container/src/main/scala/com/yahoo/container/standalone/LocalFileDb.scala
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.standalone

import java.io.File
import java.lang.reflect.Constructor
import java.nio.file.Path
import java.util
import java.util.concurrent.TimeUnit

import com.yahoo.config.FileReference
import com.yahoo.config.application.api.FileRegistry
import com.yahoo.config.application.api.FileRegistry.Entry
import com.yahoo.container.standalone.LocalFileDb._
import com.yahoo.filedistribution.fileacquirer.FileAcquirer
import com.yahoo.net.HostName

import scala.collection.JavaConverters._
import scala.collection.mutable


/**
 * FileAcquirer and FileRegistry working on a local directory.
 * @author tonytv
 */
class LocalFileDb(appPath: Path) extends FileAcquirer with FileRegistry {
  private val fileReferenceToFile = mutable.Map[FileReference, File]()

  /** *** FileAcquirer overrides *****/
  def waitFor(reference: FileReference, l: Long, timeUnit: TimeUnit): File = {
    synchronized {
      fileReferenceToFile.get(reference).getOrElse {
        throw new RuntimeException("Invalid file reference " + reference)
      }
    }
  }

  override def shutdown() {}

  /** *** FileRegistry overrides *****/
  def addFile(relativePath: String): FileReference = {
    val file = appPath.resolve(relativePath).toFile
    if (!file.exists) {
      throw new RuntimeException("The file does not exist: " + file.getPath)
    }

    val fileReference: FileReference = fileReferenceConstructor.newInstance("LocalFileDb:" + relativePath)
    fileReferenceToFile.put(fileReference, file)
    fileReference
  }

  def fileSourceHost: String =
    HostName.getLocalhost

  def allRelativePaths: java.util.Set[String] = {
    new java.util.HashSet(fileReferenceToFile.values.map(_.getPath).asJavaCollection)
  }

  override def export(): util.List[Entry] = {
    new java.util.ArrayList(fileReferenceToFile.keys.map{ (ref: FileReference) => new Entry(fileReferenceToFile.get(ref).get.getPath, ref)}.asJavaCollection)
  }

  override def addUri(uri: String): FileReference = {
    throw new RuntimeException("addUri(uri: String) is not implemented here.");
  }
}

object LocalFileDb {
  private def createFileReferenceConstructor: Constructor[FileReference] = {
    val method: Constructor[FileReference] = classOf[FileReference].getDeclaredConstructor(classOf[String])
    method.setAccessible(true)
    method
  }

  private val fileReferenceConstructor: Constructor[FileReference] = createFileReferenceConstructor
}

File: standalone-container/src/main/java/com/yahoo/container/standalone/LocalFileDb.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.standalone;

import com.yahoo.config.FileReference;
import com.yahoo.config.application.api.FileRegistry;
import com.yahoo.filedistribution.fileacquirer.FileAcquirer;
import net.jpountz.lz4.LZ4FrameOutputStream;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

/**
 * FileAcquirer and FileRegistry working on a local directory.
 *
 * @author Tony Vaagenes
 * @author ollivir
 */
public class LocalFileDb implements FileAcquirer, FileRegistry {

    private final Map<FileReference, File> fileReferenceToFile = new HashMap<>();
    private final Path appPath;

    public LocalFileDb(Path appPath) {
        this.appPath = appPath;
    }

    /* FileAcquirer overrides */
    @Override
    public File waitFor(FileReference reference, long l, TimeUnit timeUnit) {
        synchronized (this) {
            File file = fileReferenceToFile.get(reference);
            if (file == null) {
                return new File(reference.value()); // Downloaded file reference: Will (hopefully) be resolved client side
            }
            return file;
        }
    }

    @Override
    public void shutdown() {
    }

    /* FileRegistry overrides */
    @Override
    public FileReference addFile(String relativePath) {
        File file = appPath.resolve(relativePath).toFile();
        Path relative = appPath.relativize(file.toPath()).normalize();
        if (relative.isAbsolute() || relative.startsWith(".."))
            throw new IllegalArgumentException(file + " is not a descendant of " + appPath);

        if (!file.exists()) {
            throw new RuntimeException("The file does not exist: " + file.getPath());
        }

        FileReference fileReference = new FileReference("LocalFileDb:" + relativePath);
        fileReferenceToFile.put(fileReference, file);
        return fileReference;
    }

    @Override
    public List<Entry> export() {
        return fileReferenceToFile.entrySet().stream().map(entry -> new Entry(entry.getValue().getPath(), entry.getKey()))
                .collect(Collectors.toList());
    }

    @Override
    public FileReference addUri(String uri) {
        throw new RuntimeException("addUri(String uri) is not implemented here.");
    }

    @Override
    public FileReference addBlob(String name, ByteBuffer blob) {
        writeBlob(blob, name);
        File file = appPath.resolve(name).toFile();
        FileReference fileReference = new FileReference("LocalFileDb:" + name);
        fileReferenceToFile.put(fileReference, file);
        return fileReference;
    }

    private void writeBlob(ByteBuffer blob, String relativePath) {
        try (FileOutputStream fos = new FileOutputStream(new File(appPath.toFile(), relativePath))) {
            if (relativePath.endsWith(".lz4")) {
                LZ4FrameOutputStream lz4 = new LZ4FrameOutputStream(fos);
                lz4.write(blob.array(), blob.arrayOffset(), blob.remaining());
                lz4.close();
            } else {
                fos.write(blob.array(), blob.arrayOffset(), blob.remaining());
            }
        } catch (IOException e) {
            throw new IllegalArgumentException("Failed writing temp file", e);
        }
    }

}
